README.md
# Plotly Plots from Haskell

Generates an HTML Document with a Plotly Plot. These are _not_ bindings to anything.

For more details see the [documentation](https://augustunderground.github.io/plotly-hs).

## Example

The [basic line plot example](https://plotly.com/javascript/line-charts/#basic-line-plot)
is recreated like this:

```haskell
import Graphics.Plotly

main :: IO ()
main = plot path script
  where
    script = scatter' [] [[1, 2, 3, 4], [1, 2, 3, 4]]
                         [[10, 15, 13, 17], [16, 5, 11, 9]]
           $ defaultConfig { lineMode = LinesMarkers }
    path   = "path/to/dir/plot.html"
```

Then navigate to `path/to/dir/plot.html` with a web browser.

### More Examples

More examples can be found in `./src/Graphics/Plotly/Example.hs` the results of
which are stored in `./plots`.

## Usage with Stack

Add this to `stack.yaml`:

```yaml
extra-deps:
  - github: AugustUnderground/plotly-hs
    commit: 62951b8991d991ff46967e3dd0742bcd9c89bdac
```


CHANGELOG.md
# Changelog for `plotly-hs`

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to the
[Haskell Package Versioning Policy](https://pvp.haskell.org/).

## Unreleased

## 0.1.0.0 - YYYY-MM-DD


src/Graphics/Plotly.hs
{-# OPTIONS_GHC -Wall #-}

{-# LANGUAGE RecordWildCards #-} 
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE DuplicateRecordFields #-}

-- | Plotly Plots in Haskell
module Graphics.Plotly ( plot, scatter', scatter, histogram, heatmap
                       , scatter3d, surface'', surface', surface
                       , parcoord'', parcoord', parcoord
                       , Script, Mode (..), Type (..), Layout (..), BarMode (..)
                       , AxisType (..), Color (..), ColorMap, ColorScale (..)
                       , Symbol (..), Line (..), Margin (..), PlotConfig (..)
                       , defaultConfig
                       ) where

import           Data.List                      (zipWith4)
import qualified Data.ByteString.Lazy     as BL
import           Graphics.Plotly.Internal
import           Graphics.Plotly.Default

-- | Scatter Plot with individual x values per trace
scatter' :: [String]  -- ^ Trace Names
         -> [[Double]] -- ^ xs
         -> [[Double]] -- ^ ys
         -> PlotConfig -- ^ Plot Config
         -> Script     -- ^ Plotly Script
scatter' ns xs ys cfg@PlotConfig{..} = toScript layout traces
  where
    z'          = []
    layout      = Just $ fromConfig cfg
    traces      = if null ns then zipWith (mt Nothing) xs ys else zipWith3 mt (map Just ns) xs ys
    mt n' x' y' = mkTrace n' (Just lineMode) (Just marker) Scatter barMode bins x' y' z'

-- | Scatter Plot with same x axis for all ys
scatter :: [String]   -- ^ Trace Names
        -> [Double]   -- ^ xs
        -> [[Double]] -- ^ ys
        -> PlotConfig -- ^ Plot Config
        -> Script     -- ^ Plotly Script
scatter ns x ys cfg = scatter' ns xs ys cfg
  where
    xs = replicate (length ys) x

-- | Histogram Plot
histogram :: [String]    -- ^ Trace Names
          -> [[Double]]  -- ^ Traces
          -> PlotConfig  -- ^ Plot Config
          -> Script      -- ^ Plotly Script
histogram ns xs cfg@PlotConfig{..} = toScript layout traces
  where
    layout = Just $ fromConfig cfg
    traces = zipWith (\ns' xs' -> mkTrace (Just ns') Nothing Nothing Histogram barMode bins xs' [] []) ns xs

-- | Heatmap Plot
heatmap :: [String]   -- ^ X Categories
        -> [String]   -- ^ Y Categories
        -> [[Double]] -- ^ Data Matrix
        -> PlotConfig -- ^ Plot Config
        -> Script     -- ^ Plotly Script
heatmap xs ys zs cfg@PlotConfig{..} = toScript layout [traceh]
  where
    layout = Just $ fromConfig cfg
    traceh = mkTraceH colorScale True True xs ys zs

-- | 3D Scatter Plot
scatter3d :: [String]   -- ^ Trace Names
          -> [[Double]] -- ^ xs
          -> [[Double]] -- ^ ys
          -> [[Double]] -- ^ zs
          -> PlotConfig -- ^ Plot Config
          -> Script     -- ^ Plotly Script
scatter3d ns xs ys zs cfg@PlotConfig{..} = toScript layout traces
  where
    layout         = Just $ fromConfig cfg
    traces         = zipWith4 mt ns xs ys zs
    mt n' x' y' z' = mkTrace (Just n') (Just lineMode) (Just marker) Scatter3D barMode bins x' y' z'

-- | 3D Surface Plot
surface :: [String]     -- ^ Trace Names
        -> [[Double]]   -- ^ xs
        -> [[Double]]   -- ^ ys
        -> [[[Double]]] -- ^ zs
        -> PlotConfig   -- ^ P
        -> Script       -- ^ Trace
surface ns xs ys zs cfg@PlotConfig{..} = toScript layout traces
  where
    layout   = Just $ fromConfig cfg
    traces   = zipWith4 mt ns xs ys zs
    mt n' x' y' z' = mkTraceS (Just n') colorScale x' y' z'

-- | 3D Surface Plot with same xs and ys for all zs
surface' :: [String]     -- ^ Trace Names
         -> [Double]     -- ^ xs
         -> [Double]     -- ys
         -> [[[Double]]] -- ^ zs
         -> PlotConfig   -- ^ Plot Config
         -> Script       -- ^ Plotly Script
surface' ns xs ys = surface ns (repeat xs) (repeat ys)

-- | 3D Surface Plot where X and Y are indices
surface'' :: [String]    -- ^ Trace Names
         -> [[[Double]]] -- ^ zs
         -> PlotConfig   -- ^ Plot Config
         -> Script       -- ^ Ploty Script
surface'' ns = surface' ns [] []

-- | Parallel Coordinate Plot with line gradient
parcoord :: [Double]     -- ^ Color column
         -> [String]     -- ^ Axis Labels
         -> [[[Double]]] -- ^ traces ( rows ( values ) )
         -> PlotConfig   -- ^ Plot Config
         -> Script       -- ^ Plotly Script
parcoord cs ns vs cfg@PlotConfig{..} = toScript layout traces
  where
    layout   = Just $ fromConfig cfg
    traces   = [mkTraceP colorScale showScale reverseScale cs ns vs]

-- | Parallel Coordinate Plot with discrete line colors
parcoord' :: [Double]     -- ^ Color column
          -> [String]     -- ^ Axis Labels
          -> [[[Double]]] -- ^ traces ( rows ( values ) )
          -> PlotConfig   -- ^ Plot Config
          -> Script       -- ^ Plotly Script
parcoord' cs ns vs cfg@PlotConfig{..} = toScript layout traces
  where
    layout   = Just $ fromConfig cfg
    traces   = [mkTraceP colorScale False False cs ns vs]

-- | Parallel Coordinate Plot without line gradient
parcoord'' :: [String]     -- ^ Axis Names
           -> [[[Double]]] -- ^ traces ( rows ( values ) )
           -> PlotConfig   -- ^ Plot Config
           -> Script       -- ^ Plot Script
parcoord'' ns vs cfg = toScript layout traces
  where
    layout   = Just $ fromConfig cfg 
    traces   = [mkTraceP' ns vs]

-- | Save Plot to HTML File
plot :: FilePath -- ^ Target path to .html file
     -> Script   -- ^ Plotly Script
     -> IO ()
plot path = BL.writeFile path . toHtml


src/Graphics/Plotly/Default.hs
{-# OPTIONS_GHC -Wall #-}

{-# LANGUAGE RecordWildCards #-} 
{-# LANGUAGE DuplicateRecordFields #-}

-- | Default Plot Configuration
module Graphics.Plotly.Default ( PlotConfig (..), defaultConfig, fromConfig
                               ) where

import Graphics.Plotly.Internal

-- | Config to Layout
fromConfig :: PlotConfig -> Layout
fromConfig PlotConfig{..} = layout
  where
    xaxis  = Just $ Axis xlabel True True xmode
    yaxis  = Just $ Axis ylabel True True ymode
    barm   = Just barMode
    layout = Layout title' xaxis yaxis barm height width
                    (Just legend) (Just margin)

-- | Trace Independent Plot Config
data PlotConfig = PlotConfig { title'       :: !String     -- ^ Plot Title (default "")
                             , xlabel       :: !String     -- ^ X Axis Label (default "")
                             , ylabel       :: !String     -- ^ Y Axis Label (default "")
                             , xmode        :: !AxisType   -- ^ X Axis Type (default Linear)
                             , ymode        :: !AxisType   -- ^ Y Axis Type (default Linear)
                             , barMode      :: !BarMode    -- ^ Bar Mode (default Stack)
                             , lineMode     :: !Mode       -- ^ Line Mode (default Linear)
                             , colorScale   :: !ColorScale -- ^ Color Scale for 3D (default Viridis)
                             , showScale    :: !Bool       -- ^ Show Color Bar (default True)
                             , reverseScale :: !Bool       -- ^ Reverse the color scale (default False)
                             , legend       :: !Bool       -- ^ Show Legend (default True)
                             , height       :: !Int        -- ^ Canvas Height in Pixels (default 800)
                             , width        :: !Int        -- ^ Canvas Width in Pixels (default 800)
                             , margin       :: !Margin     -- ^ Layout margin (default 66 66 66 66)
                             , marker       :: !Marker     -- ^ Marker (default 0.5 Circle 1.0)
                             , bins         :: !XBins      -- ^ Histogram bins (default 0.05 Nothing Nothing)
                             } deriving (Show)

-- | Default @PlotConfig@
defaultConfig :: PlotConfig
defaultConfig = PlotConfig { title'       = ""
                           , xlabel       = ""
                           , ylabel       = ""
                           , xmode        = Linear
                           , ymode        = Linear
                           , barMode      = Overlay
                           , lineMode     = Lines
                           , colorScale   = Viridis
                           , showScale    = True
                           , reverseScale = False
                           , legend       = True
                           , height       = 800
                           , width        = 800
                           , margin       = Margin 66 66 66 66
                           , marker       = Marker 0.1 Circle 1.0
                           , bins         = XBins 0.05 Nothing Nothing
                           }


src/Graphics/Plotly/Internal.hs
{-# OPTIONS_GHC -Wall -fno-warn-missing-export-lists #-}

{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE DeriveAnyClass #-}
{-# LANGUAGE RecordWildCards #-}
{-# LANGUAGE TypeApplications #-}
{-# LANGUAGE OverloadedStrings #-} 
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE DuplicateRecordFields #-}

-- | Internal Functions and Types for Plotly Plots in Haskell
module Graphics.Plotly.Internal where

import           Data.List                        (nub)
import           Data.Aeson
import           Data.Aeson.Types
import           GHC.Generics
import qualified Data.ByteString.Lazy       as BL
import qualified Data.ByteString.Lazy.Char8 as C8
import           Data.ByteString.Lazy             (ByteString)

-- | Thanks Alan Zimmerman
-- https://gist.github.com/alanz/2465584
omitNulls :: [Pair] -> [Pair]
omitNulls = filter ((/= Null) . snd)

-- | Prepend Color Column to Parallel Coordinate Plot
prependColor :: [[Double]] -> Int -> [[Double]]
prependColor ds i = is : ds
  where
    is = replicate (length $ head ds) $ realToFrac i

-- | Internal Type Alias
type Script = ByteString

-- | Takes a Script and inserts it into the default HTML template
toHtml :: Script -> ByteString
toHtml script = C8.unlines [ "<head>"
                           ,   "<script src=\"https://cdn.plot.ly/plotly-latest.min.js\"></script>"
                           , "</head>"
                           , "<body>"
                           ,   "<div id=\"plotDiv\"></div>"
                           ,   "<script>", script, "</script>"
                           , "</body>" ]

-- | Layout Margins
data Margin = Margin { l :: !Int -- ^ Left Margin
                     , r :: !Int -- ^ Right Margin
                     , b :: !Int -- ^ Bottom Margin
                     , t :: !Int -- ^ Top Margin
                     } deriving (Eq, Show, Generic, ToJSON)

-- | Layout for Plots
data Layout = Layout { title      :: !String          -- ^ Plot Title
                     , xaxis      :: !(Maybe Axis)    -- ^ X Axis Layout
                     , yaxis      :: !(Maybe Axis)    -- ^ Y Axis Layout
                     , barmode    :: !(Maybe BarMode) -- ^ Bar Chart Layout
                     , height     :: !Int             -- ^ Plot Height in Pixels
                     , width      :: !Int             -- ^ Plot Width in Pixels
                     , showlegend :: !(Maybe Bool)    -- ^ Show a Legend
                     , margin     :: !(Maybe Margin)  -- ^ Margins
                     } deriving (Show, Generic, ToJSON)

-- | Axis Type
data AxisType = Linear          -- ^ Linear Axis (default)
              | Log             -- ^ Logarithmic Axis
              | Date            -- ^ Date Axis
              | Category        -- ^ Categorical Axis
              | MultiCategory   -- ^ Multi Category Axis (?)
              deriving (Eq, Show, Generic, ToJSON)

-- | Axis Layout
data Axis = Axis { title    :: !String   -- ^ Axis Label
                 , showGrid :: !Bool     -- ^ Enable Grid
                 , zeroLine :: !Bool     -- ^ Enable Zero Line
                 , type'    :: !AxisType -- ^ Axis Type
                 } deriving (Show, Generic)

instance ToJSON Axis where
  toJSON Axis{..} = object [ "title"    .= title
                           , "showgrid" .= showGrid
                           , "zeroline" .= zeroLine
                           , "type"     .= type' ]

-- | Parallel Coordinate Lines
data Line = Line { showscale    :: !(Maybe Bool)       -- ^ Show Color Bar
                 , reversescale :: !(Maybe Bool)       -- ^ Reverse the Color Scale
                 , colorscale   :: !(Maybe ColorScale) -- ^ Color Scale
                 , colormap     :: !(Maybe ColorMap)   -- ^ Color Map
                 , color        :: !(Maybe [Double])   -- ^ Values represnting the color
                 } deriving (Show, Generic)

instance ToJSON Line where
  toJSON Line{..} = object
                  $ omitNulls [ "showscale"    .= showscale
                              , "color"        .= color
                              , "reversescale" .= reversescale
                              , "colorscale"   .= colormap
                              , "colorscale"   .= colorscale ]

-- | Marker Config
data Marker = Marker { size    :: !Double -- ^ Marker Size
                     , symbol  :: !Symbol -- ^ Marker Symbol
                     , opacity :: !Double -- ^ Marker Opacity
                     } deriving (Show, Generic, ToJSON)

-- | Data Trace
data Trace = Trace { name    :: !(Maybe String)    -- ^ Trace Name
                   , x       :: !(Maybe [Double])  -- ^ X axis values
                   , y       :: !(Maybe [Double])  -- ^ Y axis values
                   , z       :: !(Maybe [Double])  -- ^ Z axis values
                   , mode    :: !(Maybe Mode)      -- ^ Scatter Mode
                   , type'   :: !Type              -- ^ Plot Type
                   , barmode :: !(Maybe BarMode)   -- ^ Histogram Bar Mode
                   , xbins   :: !(Maybe XBins)     -- ^ Histogram Bins 
                   , marker  :: !(Maybe Marker)    -- ^ Marker Configuration
                   } deriving (Generic, Show)

instance ToJSON Trace where
  toJSON Trace{..} = object
                   $ omitNulls [ "name"    .= name
                               , "x"       .= x
                               , "y"       .= y
                               , "z"       .= z
                               , "mode"    .= mode
                               , "type"    .= type'
                               , "marker"  .= marker
                               , "barmode" .= barmode 
                               , "xbins"   .= xbins ]

-- | Heatmap Trace
data TraceH = TraceH { z           :: ![[Double]]         -- ^ Data Matrix
                     , x           :: !(Maybe [String])   -- ^ Y labels
                     , y           :: !(Maybe [String])   -- ^ Y Labels
                     , type'       :: !Type               -- ^ Heatmap
                     , hoverOnGaps :: !(Maybe Bool)       -- ^ Hover on Gaps
                     , showScale   :: !(Maybe Bool)       -- ^ Show scale Bar
                     , colorScale  :: !(Maybe ColorScale) -- ^ Color Scale
                     } deriving (Show)

instance ToJSON TraceH where
  toJSON TraceH{..} = object [ "z"           .= z
                             , "x"           .= x
                             , "y"           .= y
                             , "hoverongaps" .= hoverOnGaps
                             , "showscale"   .= showScale
                             , "colorscale"  .= colorScale
                             , "type"        .= type' ]

-- | Surface Trace
data TraceS = TraceS { name        :: !(Maybe String)     -- ^ Trace Name
                     , z           :: ![[Double]]         -- ^ Data Matrix
                     , x           :: !(Maybe [Double])   -- ^ Y labels
                     , y           :: !(Maybe [Double])   -- ^ Y Labels
                     , type'       :: !Type               -- ^ Surface
                     , colorScale  :: !(Maybe ColorScale) -- ^ Color Scale
                     } deriving (Show)

instance ToJSON TraceS where
  toJSON TraceS{..} = object [ "name"       .= name
                             , "z"          .= z
                             , "x"          .= x
                             , "y"          .= y
                             , "colorscale" .= colorScale
                             , "type"       .= type' ]

-- | Parallel Coordinate Trace
data TraceP = TraceP { type'      :: !Type         -- ^ ParCoords
                     , line       :: !(Maybe Line) -- ^ Line Configuration
                     , dimensions :: ![Dimension]  -- ^ Dimensions
                     } deriving (Show)

instance ToJSON TraceP where
  toJSON TraceP{..} = object
                    $ omitNulls [ "dimensions" .= dimensions
                                , "line"       .= line
                                , "type"       .= type' ]

-- | Axis Dimension for Parallel Coordinate Plot
data Dimension = Dimension { range  :: !(Maybe [Int])
                           , label  :: !String
                           , values :: ![Double]
                           } deriving (Show, Generic, ToJSON)

-- | Plot Type
data Type = Scatter   -- ^ Scatter Plot
          | Histogram -- ^ Histogram
          | Heatmap   -- ^ Heatmap
          | Scatter3D -- ^ 3D Scatter
          | Surface   -- ^ 3D Surface
          | ParCoords -- ^ Parallel Coordinate Plot
          deriving (Eq, Generic, Show)

instance ToJSON Type where
  toJSON Scatter   = "scatter"
  toJSON Histogram = "histogram"
  toJSON Heatmap   = "heatmap"
  toJSON Scatter3D = "scatter3d"
  toJSON Surface   = "surface"
  toJSON ParCoords = "parcoords"

-- | Scatter Mode
data Mode = Lines        -- ^ Draw Lines
          | Markers      -- ^ Draw Markers only
          | LinesMarkers -- ^ Draw Both
          deriving (Eq, Generic, Show)

instance ToJSON Mode where
  toJSON Lines        = "lines"
  toJSON Markers      = "markers"
  toJSON LinesMarkers = "lines+markers"

-- | Marker Symbols
data Symbol = Circle
            | CircleOpen
            | Cross
            | Diamond
            | DiamondOpen
            | Square
            | SquareOpen
            | X
            deriving (Eq, Show, Generic)

instance ToJSON Symbol where
  toJSON Circle      = "circle"
  toJSON CircleOpen  = "circle-open"
  toJSON Cross       = "cross"
  toJSON Diamond     = "diamond"
  toJSON DiamondOpen = "diamond-open"
  toJSON Square      = "square"
  toJSON SquareOpen  = "square-open"
  toJSON X           = "x"

-- | Color scales for Heamap / 3D Plots
data ColorScale = Discrete
                | Blackbody
                | Bluered
                | Blues
                | Cividis
                | Earth
                | Electric
                | Greens
                | Greys
                | Hot
                | Jet
                | Picnic
                | Portland
                | Rainbow
                | RdBu
                | Reds
                | Viridis
                | YlGnBu
                | YlOrRd
  deriving (Eq, Show, Generic, ToJSON)

-- | Some Color Strings
data Color = Red
           | Green
           | Blue
           | Yellow
           | Magenta
           | Cyan
           | Black
           | Gray
           | White
            deriving (Enum, Eq, Generic)

instance Show Color where
  show Red     = "red"
  show Green   = "green"
  show Blue    = "blue"
  show Yellow  = "yellow"
  show Magenta = "magenta"
  show Cyan    = "cyan"
  show Black   = "black"
  show Gray    = "gray"
  show White   = "white"

instance ToJSON Color where
  toJSON Red     = "red"
  toJSON Green   = "green"
  toJSON Blue    = "blue"
  toJSON Yellow  = "yellow"
  toJSON Magenta = "magenta"
  toJSON Cyan    = "cyan"
  toJSON Black   = "black"
  toJSON Gray    = "gray"
  toJSON White   = "white"

-- | Type alias for mapping a color
data ColorMapping = ColorMapping !Double !Color
    deriving (Generic, ToJSON)

instance Show ColorMapping where
  show (ColorMapping n c) = "[" ++ show n ++ ", \"" ++ show c ++ "\"]"

-- | Type alias for a list of color mappings
type ColorMap = [ColorMapping]

-- | Bar Layout
data BarMode = Stack    -- ^ Stacked
             | Group    -- ^ Grouped
             | Overlay  -- ^ Overlayed
             | Relative -- ^ Relative
             deriving (Eq, Generic, Show)

instance ToJSON BarMode where
  toJSON Stack    = "stack"
  toJSON Group    = "group"
  toJSON Overlay  = "overlay"
  toJSON Relative = "relative"

-- | Bins in Histogram
data XBins = XBins { size  :: !Double           -- ^ Size of Bins
                   , start :: !(Maybe Double)   -- ^ Start of Bins
                   , end   :: !(Maybe Double)   -- ^ End of Bins
                   } deriving (Show, Generic, ToJSON)

-- | Generates line for @data@ variable for @N@ traces
traceData :: Int -> Script
traceData num = BL.concat [ "var data = [", ids, "];" ]
  where
    ids  = C8.intercalate "," [ C8.pack $ "trace" ++ show i | i <- [0 .. num - 1] ]

-- | Generates line for @i@th @trace@ variable
traceLine :: (ToJSON a) => a -> Int -> ByteString
traceLine t i = BL.concat ["var trace", i', " = ", encode t, ";"]
  where
    i' = C8.pack (show i)

-- | Converts layout and list of traces to @Script@ for insertion into HTML
toScript :: (ToJSON a) => Maybe Layout -> [a] -> ByteString
toScript layout traces = C8.unlines $ ls ++ [ds, lay, "Plotly.newPlot('plotDiv', data, layout);"]
  where
    num = length traces
    ls  = zipWith traceLine traces [ 0 .. ]
    ds  = traceData num
    lay = BL.concat ["var layout = ", encode layout, ";"]

-- | Trace contructor, different order of arguments
mkTrace :: Maybe String -> Maybe Mode -> Maybe Marker -> Type -> BarMode
        -> XBins -> [Double] -> [Double] -> [Double] -> Trace
mkTrace n m m' t b s xs ys zs = Trace n xs' ys' zs' m t bm' bs' m'
  where
    bs' = if t == Histogram then Just s else Nothing
    bm' = if t == Histogram then Just b else Nothing
    xs' = if null xs then Nothing else Just xs
    ys' = if null ys then Nothing else Just ys
    zs' = if null zs then Nothing else Just zs

-- | Heatmap Trace constructor
mkTraceH :: ColorScale -> Bool -> Bool -> [String] -> [String] -> [[Double]] -> TraceH
mkTraceH cs hv sc xs ys zs = TraceH zs xs' ys' Heatmap (Just hv) (Just sc) (Just cs)
  where
    ys' = if null ys then Nothing else Just ys
    xs' = if null xs then Nothing else Just xs

-- | Heatmap Trace constructor
mkTraceS :: Maybe String -> ColorScale -> [Double] -> [Double] -> [[Double]] -> TraceS
mkTraceS ns cs xs ys zs = TraceS ns zs xs' ys' Surface (Just cs)
  where
    ys' = if null ys then Nothing else Just ys
    xs' = if null xs then Nothing else Just xs

-- | Parallel Coordinate Trace Constructor
mkTraceP :: ColorScale -> Bool -> Bool -> [Double] -> [String] -> [[[Double]]] -> TraceP
mkTraceP scale show' rev' colors labels ds = TraceP ParCoords (Just line) dims
  where
    mapping = zipWith ColorMapping (nub colors) [ Red .. ]
    line = if scale == Discrete then Line (Just show') (Just rev') Nothing (Just mapping) (Just colors)
                                else Line (Just show') (Just rev') (Just scale) Nothing (Just colors)
    ds'  = foldl1 (zipWith (++)) ds
    dims = zipWith (Dimension Nothing) labels ds'

-- | Parallel Coordinate Trace Constructor without Line configuration
mkTraceP' :: [String] -> [[[Double]]] -> TraceP
mkTraceP' labels ds = TraceP ParCoords (Just line) dims
  where
    num    = realToFrac $ length ds
    len    = length . head $ head ds
    colors = if length ds > 1 then Just . foldl1 (++) $ map (replicate len) [1.0 .. num]
                              else Nothing
    line   = Line (Just False) (Just False) Nothing Nothing colors
    ds'    = foldl1 (zipWith (++)) ds
    dims   = zipWith (Dimension Nothing) labels ds'


src/Graphics/Plotly/Example.hs
{-# OPTIONS_GHC -Wall -fno-warn-missing-export-lists #-}

{-# LANGUAGE DuplicateRecordFields #-}

-- | Examples for using plotly-hs
module Graphics.Plotly.Example where

import Control.Monad
import System.Random
import Graphics.Plotly

-- | Put All plots in the plots directory
plotPath :: String -> FilePath
plotPath name = "./plots/" ++ name ++ ".html"

-- | 2D Scatter Plot as seen here:
-- https://plotly.com/javascript/line-charts/#basic-line-plot
scatterExample :: IO ()
scatterExample = plot (plotPath "scatter") scatterPlot
  where
    scatterPlot = scatter' ["foo", "bar"]                     -- Trace Names
                           [[1, 2, 3, 4], [1, 2, 3, 4]]       -- X values
                           [[10, 15, 13, 17], [16, 5, 11, 9]] -- Y Values
                $ defaultConfig { lineMode = LinesMarkers }   -- Lines and Markers

-- | Histogram / Bar chart Example
histExample :: IO ()
histExample = do
    x1 <- replicateM 100 (randomRIO (0, 5) ) :: IO [Double]
    x2 <- replicateM 100 (randomRIO (0, 5) ) :: IO [Double]
    x3 <- replicateM 100 (randomRIO (0, 5) ) :: IO [Double]
    let histPlot = histogram ["x1", "x2", "x3"] [x1,x2,x3]
                 $ defaultConfig {title' = "A Histogram of Random Data"}
    plot (plotPath "hist") histPlot

-- Heatmap Plot as seen here:
-- https://plotly.com/javascript/heatmaps/#heatmap-with-categorical-axis-labels
heatmapExample :: IO ()
heatmapExample = plot (plotPath "heatmap") heatPlot
  where
    heatPlot = heatmap ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday"]
                       ["Morning", "Afternoon", "Evening"]
                       [[1, 0, 30, 50, 1], [20, 1, 60, 80, 30], [30, 60, 1, -10, 20]]
                       defaultConfig

-- | 3D Scatter Plot Example
scatter3dExample :: IO ()
scatter3dExample = plot (plotPath "scatter3d") scatter3dPlot
  where
    scatter3dPlot = scatter3d ["point cloud"] [[1, 0, 30, 50, 1]] 
                              [[20, 1, 60, 80, 30]] [[30, 60, 1, -10, 20]]
                  $ defaultConfig {lineMode = Markers}

-- | 3D Surface Plot Example
surfaceExample :: IO ()
surfaceExample = plot (plotPath "surface1") surfPlot''
              >> plot (plotPath "surface2") surfPlot'
  where
    xx = map ((*10) . subtract 5) [0 .. 14]
    yy = [2,4,6,8,10,12]
    zz = [ [8.83,8.89,8.81,8.87,8.9,8.87]
         , [8.89,8.94,8.85,8.94,8.96,8.92]
         , [8.84,8.9,8.82,8.92,8.93,8.91]
         , [8.79,8.85,8.79,8.9,8.94,8.92]
         , [8.79,8.88,8.81,8.9,8.95,8.92]
         , [8.8,8.82,8.78,8.91,8.94,8.92]
         , [8.75,8.78,8.77,8.91,8.95,8.92]
         , [8.8,8.8,8.77,8.91,8.95,8.94]
         , [8.74,8.81,8.76,8.93,8.98,8.99]
         , [8.89,8.99,8.92,9.1,9.13,9.11]
         , [8.97,8.97,8.91,9.09,9.11,9.11]
         , [9.04,9.08,9.05,9.25,9.28,9.27]
         , [9,9.01,9,9.2,9.23,9.2]
         , [8.99,8.99,8.98,9.18,9.2,9.19]
         , [8.93,8.97,8.97,9.18,9.2,9.18] ]
    surfPlot'' = surface'' ["surface"] [zz] defaultConfig
    surfPlot'  = surface'  ["surface"] xx yy [zz] defaultConfig


-- | Parallel Coordinate Plot Example
parcoordsExample :: IO ()
parcoordsExample = plot (plotPath "parcoords1") parPlot'
                >> plot (plotPath "parcoords2") parPlot
  where
    as = [1,2,3,5]
    bs = [3.1,1.5,4.5,9.3]
    cs = [2,4,6,8]
    ds = [-3,-5,-10,-7]
    c' = [0.0,1.0,0.0,0.0,1.0,1.0,0.0,1.0]
    parPlot' = parcoord'' ["A", "B", "C", "D"] [[as,bs,cs,ds], [bs,ds,as,cs]]
             $ defaultConfig {width = 1500, height = 600}
    parPlot  = parcoord' c' ["A", "B", "C", "D"] [[as,bs,cs,ds], [bs,ds,as,cs]]
             $ defaultConfig {width = 1500, height = 600, colorScale = Discrete}
